lightclient/src/rpc_to_internal.rs

```diff
@@ -0,0 +1,107 @@
+// Copyright 2019-2024 ChainSafe Systems
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+use anyhow::Context;
+use base64::{Engine as _, engine::general_purpose};
+use filecoin_f3_certs as certs;
+use filecoin_f3_gpbft::chain;
+use filecoin_f3_gpbft::{BitField, Cid, ECChain, PubKey, SupplementalData};
+use filecoin_f3_rpc::{self as rpc};
+use keccak_hash::H256;
+
+/// Converts RPC power entry to the internal format
+pub fn convert_power_entry(val: rpc::PowerEntry) -> anyhow::Result<filecoin_f3_gpbft::PowerEntry> {
+    Ok(filecoin_f3_gpbft::PowerEntry {
+        id: val.id,
+        power: val.power,
+        pub_key: PubKey(general_purpose::STANDARD.decode(&val.pub_key)?), // base64 -> bytes -> PubKey
+    })
+}
+
+/// Converts RPC finality certificate to the internal format
+pub fn convert_certificate(
+    val: rpc::FinalityCertificate,
+) -> anyhow::Result<certs::FinalityCertificate> {
+    Ok(certs::FinalityCertificate {
+        gpbft_instance: val.instance,
+        ec_chain: convert_chain(val.ec_chain)?,
+        supplemental_data: SupplementalData {
+            commitments: Default::default(),
+            power_table: Cid::try_from(val.supplemental_data.power_table.cid.as_str())?,
+        },
+        power_table_delta: val
+            .power_table_delta
+            .into_iter()
+            .map(convert_power_table_delta)
+            .collect::<anyhow::Result<Vec<_>, _>>()?,
+        signers: convert_bitfield(val.signers)?,
+        signature: general_purpose::STANDARD.decode(&val.signature)?, // base64 -> bytes
+    })
+}
+
+fn convert_chain(val: Vec<rpc::ECTipSet>) -> anyhow::Result<ECChain> {
+    if val.is_empty() {
+        return Err(anyhow::anyhow!("empty ec_chain"));
+    }
+
+    let tipsets = val
+        .into_iter()
+        .map(convert_tipset)
+        .collect::<anyhow::Result<Vec<_>, _>>();
+    let tipsets = tipsets?;
+
+    let base_native = tipsets[0].clone();
+    let suffix = tipsets[1..].to_vec();
+
+    ECChain::new(base_native, suffix).context("failed to create ECChain")
+}
+
+fn convert_tipset(val: rpc::ECTipSet) -> anyhow::Result<certs::Tipset> {
+    Ok(certs::Tipset {
+        epoch: val.epoch as i64,
+        key: convert_tipset_key(&val.key)?,
+        power_table: Cid::try_from(val.power_table.cid)?,
+        commitments: H256::from_slice(&general_purpose::STANDARD.decode(&val.commitments)?), // base64 -> bytes -> H256
+    })
+}
+
+fn convert_tipset_key(cid_refs: &[rpc::CidRef]) -> anyhow::Result<chain::TipsetKey> {
+    let mut key = Vec::new();
+    for cid_ref in cid_refs {
+        let cid = Cid::try_from(cid_ref.cid.as_str())?;
```

Comment:
Hm, do we really need to convert to a string? Seems wasteful. Not a huge deal, no need to premature optimisation but it's weird if `Cid` wouldn't have byte constructors.

Reply:
It's up to the F3 API to optimize this. It currently sends base32 strings.